### ISSUE TITLE
Change EmojiTrie implementation to SuffixTree 

### ISSUE THREAD

**Issue by hannojg** [Expensify/App#51148]
For this issue ticket we have implemented a very efficient/fast suffix tree implementation:

- https://github.com/Expensify/App/issues/46591

PR:

- https://github.com/Expensify/App/pull/48652

It was discussed that we want to use the implementation of the SuffixTree for the emoji search as well:

- [Discussion here](https://margelo.slack.com/archives/C05LX9D6E07/p1725560880635709?thread_ts=1723209631.007999&cid=C05LX9D6E07)

This enables:

- more efficient search in emojis
- faster construction of the emoji tree (we should compare the construction time before and after the change.)
- searching for any substring of an emoji name (not just perfect name matches)

**Comment by hannojg** [comment]
cc @roryabraham

**Comment by ChavdaSachin** [comment]
Hey @hannojg, I have been very passionate Competitive Programmer back during my college days.
I think I could be a help here.
Do you think there is a scope of an External contributor here?

**Comment by hannojg** [comment]
I think this is more of an internal ticket. The suffix tree implementation has already been built by our internal team and this is a follow up issue from one of our PRs.

**Comment by ChavdaSachin** [comment]
Alright, 
But I have one more suggestion here, we could make use of c++ here instead of js , c++ is well known for it's speed, I am really positive that c++ would make significant improvement here.

**Comment by ChavdaSachin** [comment]
Alright, 
But I have one more suggestion here, we could make use of c++ here instead of js , c++ is well known for it's speed, I am really positive that c++ would make significant improvement here.

**Comment by hannojg** [comment]
@mountiny can you please reopen? The initial PR for changing the search algorithm to a search tree took longer than anticipated. We are going to look into these items next, so its still l a priority!

**Comment by hannojg** [comment]
> But I have one more suggestion here, we could make use of c++ here instead of js , c++ is well known for it's speed, I am really positive that c++ would make significant improvement here.

Yes, while c++ might be faster there is a significant cost for bringing the data from JS over to C++ which might eat away the performance benefits. The current typescript implementation is already pretty fast.
We could open a follow up ticket to see if there is any benefit in bringing this to C++, but for now we are good I'd say

**Comment by mountiny** [comment]
@hannojg How is this looking? Could we get someone to work on this or just close it? seems like this middle ground is not very useful

**Comment by roryabraham** [comment]
Let's reopen this and get it done. Currently having two Trie implementations feels complex, plus the SuffixTree provides better matches
